packages/evm/src/eof/verify.ts

```diff
@@ -0,0 +1,519 @@
+          }
+
+          if (minStackCurrent < targetInputs) {
+            validationError(EOFError.StackUnderflow)
+          }
+
+          if (
+            maxStackCurrent + container.body.typeSections[target].maxStackHeight - targetInputs >
+            1024
+          ) {
+            validationError(EOFError.StackOverflow)
+          }
+
+          minStackNext += targetOutputs - targetInputs
+          maxStackNext += targetOutputs - targetInputs
+        } else {
+          // JUMPF
+          const currentOutputs = container.body.typeSections[codeSection].outputs
+          const targetOutputs = container.body.typeSections[target].outputs
+          const targetInputs = container.body.typeSections[target].inputs
+          const targetNonReturning = targetOutputs === 0x80
+
+          if (targetOutputs > currentOutputs && !targetNonReturning) {
+            // Spec rule:
+            // JUMPF operand must point to a code section with equal or fewer number of outputs as
+            // the section in which it resides, or to a section with 0x80 as outputs (non-returning)
+            validationError(EOFError.InvalidJUMPF)
+          }
+
+          if (returningFunction && targetOutputs <= 0x7f) {
+            // Current function is returning, but target is not, cannot jump into this
+            validationError(EOFError.InvalidReturningSection)
+          }
+
+          if (targetNonReturning) {
+            // Target is returning
+            if (minStackCurrent < targetInputs) {
+              validationError(EOFError.StackUnderflow)
+            }
+          } else {
+            // Target is returning
+            const expectedStack = currentOutputs + targetInputs - targetOutputs
+            if (!(minStackCurrent === maxStackCurrent && maxStackCurrent === expectedStack)) {
+              validationError(EOFError.InvalidStackHeight)
+            }
+          }
+          if (
+            maxStackCurrent + container.body.typeSections[target].maxStackHeight - targetInputs >
+            1024
+          ) {
+            //console.log(maxStackCurrent, targetOutputs, targetInputs, targetNonReturning)
+            validationError(EOFError.StackOverflow)
+          }
+        }
+      } else if (opcode === 0xe4) {
+        // RETF
+        // Stack height must match the outputs of current code section
+        const outputs = container.body.typeSections[codeSection].outputs
+        if (!(minStackCurrent === maxStackCurrent && maxStackCurrent === outputs)) {
+          validationError(EOFError.InvalidStackHeight)
+        }
+      } else if (opcode === 0xe6) {
+        // DUPN
+        const toDup = code[ptr + 1]
+        if (toDup + 1 > minStackCurrent) {
+          validationError(EOFError.StackUnderflow)
+        }
+      } else if (opcode === 0xe7) {
+        // SWAPN
+        const toSwap = code[ptr + 1]
+        // TODO: EVMONEs test wants this to be `toSwap + 2`, but that seems to be incorrect
```

Comment:
Do we keep the stack 1 based or zero based in our code?  That would inform this I think?  Guessing maybe EVMOne is 0 based?

Reply:
We want to keep the stack as-is (I think we are zero-based). There were many off-by-one bugs which I have to fix (those are often also confusing). However I think this comment is a reply on one of EVMONEs internal tests, which seemed wrong to me (it tried to swap the wrong values, at least that's what I thought)! (I will have to check back later on this)